crates/ide-assists/src/handlers/add_missing_match_arms.rs

```diff
@@ -181,11 +204,17 @@ pub(crate) fn add_missing_match_arms(acc: &mut Assists, ctx: &AssistContext<'_>)
         target_range,
         |builder| {
             let new_match_arm_list = match_arm_list.clone_for_update();
-            let missing_arms = missing_pats
-                .map(|(pat, hidden)| {
-                    (make::match_arm(iter::once(pat), None, make::ext::expr_todo()), hidden)
+
+            // having any hidden variants means that we need a catch-all arm
+            needs_catch_all_arm |= has_hidden_variants;
+
+            let missing_arms = missing_pats.filter_map(|(pat, hidden)| {
+                // filter out hidden patterns because they're handled by the catch-all arm
+                (!hidden).then(|| {
+                    make::match_arm(iter::once(pat), None, make::ext::expr_todo())
+                        .clone_for_update()
                 })
-                .map(|(it, hidden)| (it.clone_for_update(), hidden));
+            });
```

Comment:
Can you rewrite this with `filter()` and then `map()`? I think separating filtering and ast transformation makes the code more comprehensible.